lte/gateway/c/session_manager/ShardTracker.cpp

```diff
@@ -13,45 +13,43 @@
 
 #include <string>
 #include <vector>
+#include <set>
 
 #include "ShardTracker.h"
 
 namespace magma {
 
-int ShardTracker::add_ue() {
-  /*
-  * If there are no shards, add an entry for the shard
-  * representing a single UE at the new shard
-  */
-  if (ue_count_per_shard_.size() == 0) {
-    ue_count_per_shard_.push_back(1);
-    return 0;
+uint16_t ShardTracker::add_ue(const std::string imsi) {
+  if (imsis_per_shard_.empty()) {
+    imsis_per_shard_.push_back({imsi});
   }
-  /*
-  * Iterate through all existing shards, if all are full
-  * create a new shard with quantity 1, otherwise increment
-  * the quantity of UEs in the latest shard
-  */
-  for (size_t shard_id = 0; shard_id < ue_count_per_shard_.size(); shard_id++) {
-    if (ue_count_per_shard_[shard_id] < max_shard_size_) {
-      ue_count_per_shard_[shard_id]++;
+  for (size_t shard_id = 0; shard_id < imsis_per_shard_.size(); shard_id++) {
+    // If the UE is already in the shard, return the shard id. This check
+    // is meant to avoid multiple sessions for a UE being assigned duplicate
+    // entries.
+    std::set<std::string>& imsis_at_id = imsis_per_shard_[shard_id];
+    if (imsis_at_id.find(imsi) != imsis_at_id.end()) {
```

Comment:
As shown in the template we already have a way to find in O(1) if the imsi has a shard id without having to maintain a parallel list of imsis

This is inefficient, not only because you have to go through all the shards but because you will have to  go through all of them just to find out it is not there.